paychmgr/manager.go

```diff
@@ -218,14 +199,76 @@ func (pm *Manager) CheckVoucherSpendable(ctx context.Context, ch address.Address
 	return ca.checkVoucherSpendable(ctx, ch, sv, secret, proof)
 }
 
-func (pm *Manager) AddVoucher(ctx context.Context, ch address.Address, sv *paych.SignedVoucher, proof []byte, minDelta types.BigInt) (types.BigInt, error) {
+// AddVoucherOutbound adds a voucher for an outbound channel.
+// Returns an error if the channel is not already in the store.
+func (pm *Manager) AddVoucherOutbound(ctx context.Context, ch address.Address, sv *paych.SignedVoucher, proof []byte, minDelta types.BigInt) (types.BigInt, error) {
 	ca, err := pm.accessorByAddress(ch)
 	if err != nil {
 		return types.NewInt(0), err
 	}
 	return ca.addVoucher(ctx, ch, sv, proof, minDelta)
 }
 
+// AddVoucherInbound adds a voucher for an inbound channel.
+// If the channel is not in the store, fetches the channel from state (and checks that
+// the channel To address is owned by the wallet).
+func (pm *Manager) AddVoucherInbound(ctx context.Context, ch address.Address, sv *paych.SignedVoucher, proof []byte, minDelta types.BigInt) (types.BigInt, error) {
+	// Make sure channel is in store, or can be fetched from state, and that
+	// the channel To address is owned by the wallet
+	ci, err := pm.trackInboundChannel(ctx, ch)
+	if err != nil {
+		return types.BigInt{}, err
+	}
+
+	// This is an inbound channel, so To is the Control address (this node)
+	from := ci.Target
+	to := ci.Control
+	ca, err := pm.accessorByFromTo(from, to)
+	if err != nil {
+		return types.BigInt{}, err
+	}
+	return ca.addVoucher(ctx, ch, sv, proof, minDelta)
+}
+
+func (pm *Manager) trackInboundChannel(ctx context.Context, ch address.Address) (*ChannelInfo, error) {
+	// Need to take an exclusive lock here so that channel operations can't run
+	// in parallel (see channelLock)
+	pm.lk.Lock()
+	defer pm.lk.Unlock()
+
+	// Check if channel is in store
+	ci, err := pm.store.ByAddress(ch)
+	if err == nil {
+		// Channel is in store, so it's already being tracked
+		return ci, nil
+	}
+
+	// If there's an error (besides channel not in store) return err
+	if err != ErrChannelNotTracked {
+		return nil, err
+	}
+
+	// Channel is not in store, so get channel from state
+	stateCi, err := pm.sa.loadStateChannelInfo(ctx, ch, DirInbound)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check that channel To address is in wallet
+	to := stateCi.Control // Inbound channel so To addr is Control (this node)
+	has, err := pm.pchapi.WalletHas(ctx, to)
```

Comment:
This passes the integration test, so I believe it's correct to call `WalletHas()` with the channel To address, as opposed to calling it with the channel To address account's `Address` field.
@magik6k can you confirm?

Reply:
WalletHas expects a key address (`t1..`, `t3..`), so assuming it's guaranteed to be that, it's fine

Reply:
[This is the address](https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/paych/paych_state.go#L18) that will be passed to WalletHas()

ie the "To" address of the channel state
https://github.com/filecoin-project/specs-actors/blob/a73a22a48b668c2f5a1cc6a31caef03bcaec5f70/actors/builtin/paych/paych_state.go#L11-L30
```
type State struct {
	// Channel owner, who has funded the actor
	From addr.Address
	// Recipient of payouts from channel
	To addr.Address

	// Amount successfully redeemed through the payment channel, paid out on `Collect()`
	ToSend abi.TokenAmount

	// Height at which the channel can be `Collected`
	SettlingAt abi.ChainEpoch
	// Height before which the channel `ToSend` cannot be collected
	MinSettleHeight abi.ChainEpoch

	// Collections of lane states for the channel, maintained in ID order.
	LaneStates cid.Cid // AMT<LaneState>
}
```

Reply:
Looks like the actor code normalizes those to ID addresses, so we need to call StateAccountKey on those first